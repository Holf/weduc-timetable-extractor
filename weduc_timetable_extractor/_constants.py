WEDUC_SECTION_NAME = "weduc"

WEDUC_LOGIN_TIMEOUT = 5 * 60 * 1000  # Five minutes

STUDENT_SECTION_NAME_PREFIX = "student_"
