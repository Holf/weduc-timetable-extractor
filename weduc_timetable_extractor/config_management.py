import configparser
import sys
from pathlib import Path

from .constants import STUDENT_SECTION_NAME_PREFIX, WEDUC_SECTION_NAME

config = None


def get_config():

    global config

    if config == None:
        project_root = Path(__file__).resolve().parent.parent
        config_file_path = project_root / "config.ini"

        config = configparser.ConfigParser()
        config.read(config_file_path)

    return config


def get_config_option(section, option, throw_if_absent=False):
    config = get_config()

    if not config.has_option(section, option):
        if throw_if_absent:
            sys.stderr.write(
                f"\nError: There is no '{section}.{option}' present in config.ini"
            )
            sys.exit(2)
        else:
            return None

    return config.get(section, option)


def get_weduc_credentials():

    username, password = get_config_option(
        WEDUC_SECTION_NAME, "username"
    ), get_config_option(WEDUC_SECTION_NAME, "password")

    credentials = {
        "username": username,
        "password": password,
        "credentials_present": username != None and password != None,
    }

    return credentials


def get_chromium_path():
    return get_config_option(WEDUC_SECTION_NAME, "chromium_path")


def get_student_configs():
    section_names = get_config().sections()
    student_section_names = [
        section_name.lower()
        for section_name in section_names
        if section_name.startswith(STUDENT_SECTION_NAME_PREFIX)
    ]

    student_configs = [
        get_student_config(section_name) for section_name in student_section_names
    ]

    return student_configs


def get_student_config(section_name):
    return {
        "section_name": section_name,
        "school_name": get_config_option(
            section_name, "school_name", throw_if_absent=True
        ),
        "student_name": get_config_option(
            section_name, "student_name", throw_if_absent=True
        ),
        "calendar_to_update": get_config_option(section_name, "calendar_to_update"),
    }
