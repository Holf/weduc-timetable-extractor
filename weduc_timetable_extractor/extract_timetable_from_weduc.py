import sys

from playwright.sync_api import Page

from .get_config import get_config

login_timeout = 5 * 60 * 1000  # 5 minutes


def check_config(key):
    if not (value := get_config("weduc", key)):
        sys.stderr.write(f"\nError: There is no 'weduc.{key}' present in config.ini")
        sys.exit(2)
    return value


def login(page: Page):

    if username := get_config("weduc", "username"):
        username_input = page.get_by_label("Login or E-mail")
        username_input.wait_for()
        username_input.type(username)

    if password := get_config("weduc", "password"):
        password_input = page.get_by_label("Password")
        password_input.wait_for()
        password_input.type(password)

    if username != "" and password != "":
        login_button = page.get_by_role("button", name="Login")
        login_button.click()


def extract_timetable_from_weduc(page: Page):

    school_id = check_config("school_id")
    student_id = check_config("student_id")

    login(page)

    page.wait_for_function(
        "localStorage.getItem('wfx_unq') !== null", timeout=login_timeout
    )

    page.wait_for_function(
        "localStorage.getItem('wfx_UUID') !== null", timeout=login_timeout
    )

    response_data = page.evaluate(
        """
    async ({school_id, student_id}) => {
        async function fetchData() {
            await fetch('https://app.weduc.co.uk/dashboard/index/setcurrent', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                },
                body: new URLSearchParams({
                    'entity': school_id
                })
            });
            
            const response = await fetch(`https://app.weduc.co.uk/user/profile/getTimetable/user/${student_id}/`);
            return await response.json();
        }
        return fetchData();
    }
    """,
        {"school_id": school_id, "student_id": student_id},
    )

    timetable = response_data["Body"]["timetable"]

    return timetable
