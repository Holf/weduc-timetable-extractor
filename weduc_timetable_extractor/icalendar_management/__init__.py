from .convert_transformed_timetable_to_ical import convert_transformed_timetable_to_ical
from .write_timetable_to_ics_file import write_timetable_to_ics_file
