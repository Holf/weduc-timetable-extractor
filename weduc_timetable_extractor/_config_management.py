import configparser
import platform
import sys
from pathlib import Path

from ._constants import STUDENT_SECTION_NAME_PREFIX, WEDUC_SECTION_NAME

config = None


def get_config():

    global config

    if config == None:
        if getattr(sys, "frozen", False):
            # We are running in a bundle (PyInstaller)
            project_root = Path(sys.executable).resolve().parent
        else:
            # We are running in a normal Python environment
            project_root = Path(__file__).resolve().parent.parent

        config_file_path = project_root / "config.ini"
        validate_config_ini_path(config_file_path)

        config = configparser.ConfigParser()
        config.read(config_file_path)

    return config


def get_config_option(section, option, throw_if_absent=False):
    config = get_config()

    # Check if the option exists
    if not config.has_option(section, option):
        value = None
    else:
        value = config.get(section, option).strip() or None

    # Handle the absence of value if throw_if_absent is True
    if value is None and throw_if_absent:
        sys.exit(f"\nError: There is no '{section}.{option}' present in config.ini")

    return value


def get_weduc_credentials():
    username, password = get_config_option(
        WEDUC_SECTION_NAME, "username"
    ), get_config_option(WEDUC_SECTION_NAME, "password")

    credentials = {
        "username": username,
        "password": password,
        "credentials_present": username != None and password != None,
    }

    return credentials


def get_ics_folder_path(throw_if_absent=False):
    return get_config_option("ical", "output_folder_path", throw_if_absent)


def get_chromium_path():
    chromium_path = get_config_option(WEDUC_SECTION_NAME, "chromium_path")

    if chromium_path != None:
        return chromium_path

    os_map = {
        "Windows": r"C:\Program Files (x86)\Google\Chrome\Application\chrome.exe",
        "Linux": "/usr/bin/google-chrome",
        "Darwin": "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
    }

    return os_map.get(platform.system())


def validate_config_ini_path(config_ini_file_path):
    if not config_ini_file_path.is_file():
        sys.exit(
            f"Error: unable to find a config.ini file at path '{config_ini_file_path}'"
        )


def validate_chromium_path(chromium_path):
    if not Path(chromium_path).is_file():
        sys.exit(f"Error: unable to find a browser at path '{chromium_path}'")


def get_student_configs():
    section_names = get_config().sections()
    student_section_names = [
        section_name.lower()
        for section_name in section_names
        if section_name.startswith(STUDENT_SECTION_NAME_PREFIX)
    ]

    student_configs = [
        get_student_config(section_name) for section_name in student_section_names
    ]

    validate_at_least_one_student_config(student_configs)

    return student_configs


def validate_at_least_one_student_config(student_configs):
    student_config_count = len(student_configs)

    if student_config_count > 0:
        return student_config_count

    sys.exit("Error: there are no student configs in config.ini")


def get_student_config(section_name):
    return {
        "section_name": section_name,
        "school_name": get_config_option(
            section_name, "school_name", throw_if_absent=True
        ),
        "student_name": get_config_option(
            section_name, "student_name", throw_if_absent=True
        ),
        "calendar_to_update": get_config_option(section_name, "calendar_to_update"),
    }
