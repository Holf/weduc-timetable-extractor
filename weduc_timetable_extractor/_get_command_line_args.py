import argparse

from ._config_management import get_ics_folder_path


def _get_command_line_args():
    parser = argparse.ArgumentParser(
        prog="Weduc Timetable Extractor",
        description="Extracts timetable information from Weduc and either: exports it as an iCalendar file; or, uses Google Calendar API to push it to a Google Calendar",
    )

    subparsers = parser.add_subparsers(
        help="Choose 'ical' or 'api'", dest="mode", required=True
    )

    subparsers.add_parser(
        "ical",
        help="Generate an iCalendar '.ics' file from the extracted Weduc timetable",
    )

    subparsers.add_parser(
        "api", help="Push the extracted Weduc timetable to Google Calendar"
    )

    args = parser.parse_args()

    if args.mode == "ical":
        print("validaty")
        get_ics_folder_path(throw_if_absent=True)

    return args
