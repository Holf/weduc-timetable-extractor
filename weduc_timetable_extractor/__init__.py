from ._get_command_line_args import _get_command_line_args
