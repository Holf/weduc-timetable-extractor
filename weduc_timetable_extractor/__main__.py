from playwright.sync_api import sync_playwright

from weduc_timetable_extractor import get_command_line_args
from weduc_timetable_extractor.config_management import (
    get_chromium_path,
    get_student_configs,
    get_weduc_credentials,
)
from weduc_timetable_extractor.google_calendar_management import (
    push_timetable_to_google_calendar,
)
from weduc_timetable_extractor.icalendar_management import write_timetable_to_ics_file
from weduc_timetable_extractor.weduc_interaction import (
    extract_schools_data_from_weduc,
    extract_students_data_from_weduc,
    extract_timetable_from_weduc,
    filter_and_transform_timetable,
    login_to_weduc,
    set_active_school_in_weduc,
)


def main():
    args = get_command_line_args()
    weduc_credentials = get_weduc_credentials()
    use_headless = weduc_credentials["credentials_present"]

    student_configs = get_student_configs()

    print(f"Found {len(student_configs)} student config(s)")

    with sync_playwright() as p:
        print(
            "All login credentials are present in config, so using a headless browser to extract data from Weduc."
            if use_headless
            else "A complete set of login credentials is not present in config, so using a headed browser to allow manual entry."
        )

        print("Launching browser ...")

        chromium_path = get_chromium_path() or "/usr/bin/google-chrome"
        browser = p.chromium.launch(
            executable_path=chromium_path, headless=use_headless
        )
        page = browser.new_page()

        page.goto("https://app.weduc.co.uk/")

        login_to_weduc(page, weduc_credentials)

        schools = extract_schools_data_from_weduc(page)

        for student_config in student_configs:

            _log_student_info(student_config)

            extract_student_timetable_from_weduc(page, schools, student_config)

            match args.mode:
                case "ical":
                    write_timetable_to_ics_file(args.output_folder_path, student_config)

                case "api":
                    push_timetable_to_google_calendar(student_config)


def extract_student_timetable_from_weduc(page, schools, student_config):
    _add_school_id_to_student_config(student_config, schools)
    set_active_school_in_weduc(page, student_config)

    students = extract_students_data_from_weduc(page, student_config["school_id"])
    _add_student_id_to_student_config(student_config, students)

    _extract_timetable_and_add_to_student_config(page, student_config)


def _log_student_info(student_config):
    print(
        f"""
Processing timetable for student:
    {student_config["student_name"]}
at school:
    {student_config["school_name"]}
"""
    )


def _add_school_id_to_student_config(student_config, schools):

    student_config["school_id"] = next(
        (
            school["school_id"]
            for school in schools
            if school["school_name"] == student_config["school_name"]
        ),
        None,
    )


def _add_student_id_to_student_config(student_config, students):

    student_config["student_id"] = next(
        (
            student["id"]
            for student in students
            if student["name"] == student_config["student_name"]
        ),
        None,
    )


def _extract_timetable_and_add_to_student_config(page, student_config):
    student_config["timetable"] = filter_and_transform_timetable(
        extract_timetable_from_weduc(page, student_config)
    )


if __name__ == "__main__":
    main()
