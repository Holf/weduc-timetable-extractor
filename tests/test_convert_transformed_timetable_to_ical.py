from freezegun import freeze_time

from weduc_timetable_extractor.icalendar_management import (
    _convert_transformed_timetable_to_ical,
)

test_timetable = list(
    [
        dict(
            {
                "Class name": "10w/En1",
                "Room name": "ENGLISH3",
                "Subject name": "English La",
                "Teacher name": "C Reynolds",
                "color": "rgba(255, 107, 00, 0.6)",
                "description": """
        Room name: ENGLISH3
        Class name: 10w/En1
        Subject name: English La
        Teacher name: C Reynolds
      """,
                "end": "2024-09-27T11:20:00+00:00",
                "id": "281475447549934",
                "popover": list(
                    [
                        "Room name: ENGLISH3",
                        "Class name: 10w/En1",
                        "Subject name: English La",
                        "Teacher name: C Reynolds",
                    ]
                ),
                "start": "2024-09-27T10:20:00+00:00",
                "summary": "English La - ENGLISH3 (10w/En1)",
                "title": "10w/En1",
            }
        ),
        dict(
            {
                "Class name": "10B/Dt1",
                "Room name": "DT7",
                "Subject name": "Des.Tech",
                "Teacher name": "S Oti-Akenten",
                "color": "rgba(255, 107, 00, 0.6)",
                "description": """
        Room name: DT7
        Class name: 10B/Dt1
        Subject name: Des.Tech
        Teacher name: S Oti-Akenten
      """,
                "end": "2024-09-27T12:20:00+00:00",
                "id": "281475447549935",
                "popover": list(
                    [
                        "Room name: DT7",
                        "Class name: 10B/Dt1",
                        "Subject name: Des.Tech",
                        "Teacher name: S Oti-Akenten",
                    ]
                ),
                "start": "2024-09-27T11:20:00+00:00",
                "summary": "Des.Tech - DT7 (10B/Dt1)",
                "title": "10B/Dt1",
            }
        ),
        dict(
            {
                "Class name": "10w/Ma2",
                "Room name": "T1",
                "Subject name": "Maths",
                "Teacher name": "V Tucker",
                "color": "rgba(255, 107, 00, 0.6)",
                "description": """
        Room name: T1
        Class name: 10w/Ma2
        Subject name: Maths
        Teacher name: V Tucker
      """,
                "end": "2024-09-27T14:00:00+00:00",
                "id": "281475447549936",
                "popover": list(
                    [
                        "Room name: T1",
                        "Class name: 10w/Ma2",
                        "Subject name: Maths",
                        "Teacher name: V Tucker",
                    ]
                ),
                "start": "2024-09-27T13:00:00+00:00",
                "summary": "Maths - T1 (10w/Ma2)",
                "title": "10w/Ma2",
            }
        ),
    ]
)


@freeze_time("2024-09-15")
def test_convert_json_to_ical(snapshot):

    result = _convert_transformed_timetable_to_ical(test_timetable)

    assert snapshot == result
