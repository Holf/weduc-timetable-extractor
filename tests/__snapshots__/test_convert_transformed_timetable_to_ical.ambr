# serializer version: 1
# name: test_convert_transformed_timetable_to_ical
  '''
  BEGIN:VCALENDAR
  BEGIN:VEVENT
  SUMMARY:English La - ENGLISH3 (10w/En1)
  DTSTART:20240927T102000Z
  DTEND:20240927T112000Z
  UID:281475447549934
  DESCRIPTION:\n        Room name: ENGLISH3\n        Class name: 10w/En1\n  
         Subject name: English La\n        Teacher name: C Reynolds\n      
  END:VEVENT
  BEGIN:VEVENT
  SUMMARY:Des.Tech - DT7 (10B/Dt1)
  DTSTART:20240927T112000Z
  DTEND:20240927T122000Z
  UID:281475447549935
  DESCRIPTION:\n        Room name: DT7\n        Class name: 10B/Dt1\n       
    Subject name: Des.Tech\n        Teacher name: S Oti-Akenten\n      
  END:VEVENT
  BEGIN:VEVENT
  SUMMARY:Maths - T1 (10w/Ma2)
  DTSTART:20240927T130000Z
  DTEND:20240927T140000Z
  UID:281475447549936
  DESCRIPTION:\n        Room name: T1\n        Class name: 10w/Ma2\n        
   Subject name: Maths\n        Teacher name: V Tucker\n      
  END:VEVENT
  END:VCALENDAR
  
  '''
# ---
